### PR TITLE
[PyTorch][JIT] Don't refcount Type singletons

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69579

This should help us avoid reference counting overhead on singleton Type subclasses without a major rewrite of the Type subsystem.

Differential Revision: [D32923880](https://our.internmc.facebook.com/intern/diff/D32923880/)

cc @ezyang @gchanan

### CODE REVIEW

**Comment by suo** [comment]
Should I review? Tests look sadge

**Comment by swolchok** [comment]
> Should I review? Tests look sadge

you can still review for broad direction. blocking on tests slows down the process.

**Comment by swolchok** [comment]
Looks like there's some kind of CUDA build failure I don't understand (can we not do explicit destructor calls in CUDA? why is this code getting built for CUDA anyway?), and I have to patch the Python bridging stuff for the new classes.

File: aten/src/ATen/core/type_ptr.h [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+
+   private:
+    void destroy() {
+      if (isSharedAndNonNull()) {
+        // Without SharedPtrWrapper, this line would read
+        // `shared_.~shared_ptr()` and nvcc would complain with
+        // "error: expected primary-expression before '>' token"
+        // referring to the "t" in "shared_ptr". SharedPtrWrapper
+        // exists to work around this compiler bug.
+        shared_.~SharedPtrWrapper();
```
Comment (swolchok) [inline aten/src/ATen/core/type_ptr.h:241]: CC @zasdfgbnm

Comment (zasdfgbnm) [reply aten/src/ATen/core/type_ptr.h]: cc @ptrblck

**Review by suo — APPROVED** [review]

**Comment by swolchok** [comment]
I've conservatively labeled this bc-breaking because it is possible for it to cause compile errors in downstream code that tried to use std::dynamic_pointer_cast on TypePtrs instead of going through the Type::cast API. I ran into one example of this internally and the fix was to just use Type::cast.

**Pull request closed without merging**